src/v/coproc/reconciliation_backend.cc

```diff
@@ -151,7 +201,99 @@ reconciliation_backend::process_update(update_t delta) {
     __builtin_unreachable();
 }
 
-ss::future<> reconciliation_backend::delete_non_replicable_partition(
+ss::future<std::error_code> reconciliation_backend::process_shutdown(
+  model::ntp ntp,
+  model::revision_id rev,
+  std::vector<model::broker_shard> new_replicas) {
+    vlog(coproclog.info, "Processing shutdown of: {}", ntp);
+
+    auto ntps = get_materialized_partitions(_topics.local(), ntp);
+    if (ntps.empty()) {
+        /// Input doesn't exist on this broker/shard
+        co_return make_error_code(errc::partition_not_exists);
+    }
+
+    auto rctxs = co_await _pacemaker.local().shutdown_partition(ntp);
+    if (rctxs.empty()) {
+        /// No coprocessors were processing on the shutdown input
+        co_return make_error_code(errc::partition_not_exists);
+    }
+
+    /// Remove from shard table so ntp is effectively deregistered
+    co_await _shard_table.invoke_on_all([ntps, rev](cluster::shard_table& st) {
+        for (const auto& ntp : ntps) {
+            st.erase(ntp, rev);
+        }
+    });
+
+    /// Finally remove log from disk
+    for (const auto& ntp : ntps) {
+        co_await _coproc_pm.local().remove(ntp);
+        vlog(coproclog.info, "Materialized log {} removed", ntp);
+    }
+
+    co_return make_error_code(errc::success);
+}
+
+ss::future<std::error_code> reconciliation_backend::process_restart(
+  model::ntp ntp, model::revision_id rev) {
+    vlog(coproclog.info, "Processing restart of: {}", ntp);
+
+    auto ntps = get_materialized_partitions(_topics.local(), ntp);
+    if (ntps.empty()) {
+        /// Input doesn't exist on this broker/shard
+        co_return make_error_code(errc::partition_not_exists);
+    }
+
+    for (const auto& ntp : ntps) {
+        vlog(coproclog.info, "Re-creating materialized log: {}", ntp);
+        co_await create_non_replicable_partition(ntp, rev);
+    }
+
+    /// Obtain all matching script_ids for the shutdown partition, they must all
+    /// be restarted
+    auto meta = co_await _sdb.invoke_on(
+      wasm::script_database_main_shard, [ntp](wasm::script_database& sdb) {
+          return grab_metadata(sdb, model::topic_namespace_view{ntp});
+      });
+    if (meta.empty()) {
```

Comment:
I think the best solution would be to process all coprocessor add/remove events before starting the reconciliation backends loop. This means if the condition above (`if(meta.empty()`) has been reached, it is truly an event that can be ignored, it is guaranteed that the script was eventually deleted, instead of not knowing if the script has or has not been deleted. Which would mean to log the event and "wait" for the coprocessor to be processed off of the coprocessor internal topic. Which is what this source at the time of writing this comment is doing.